File: scripts/compare_runs.py
#!/usr/bin/env python3
"""Old/new comparison of two cuMES runs (no vmecpp reference needed).

Given two run logs and two converged-state binaries, verifies that a new run
reproduces the baseline run:

  1. Converged state (the schema-v1 binary container, on-disk versions 1-8,
     or the legacy raw baseline dump): six spectral families
     (rmncc/zmnsc/lmnsc/rmnss/zmncs/lmncs), axis row (j=0) skipped, PASS at
     <= --tol relative (default 1e-8). Mirrors scripts/compare_states.py.
  2. Per-iteration residual rows parsed from the solver log
     ("%5d | %11.3e %11.3e %11.3e | %8.2e"): fsqr/fsqz/fsql/delt compared at
     <= --log-tol relative (default 5e-3). The printed precision is 3-4
     significant digits (%.3e / %.2e), so the log tolerance must be looser
     than the state tolerance; a real trajectory divergence compounds far
     beyond it within a few iterations.
  3. Restart markers: BAD JACOBIAN / BAD PROGRESS / CONVERGENCE PROBLEM
     (RESETTING DELT) must occur at the same iterations in the same order.
  4. Convergence: both runs must converge, at iterations differing by at most
     --max-iter-delta (default 10), with matching final FSQR/FSQZ/FSQL
     (within --log-tol).

Usage:
  compare_runs.py <log_a> <state_a> <log_b> <state_b> [--tol 1e-8]
                  [--log-tol 5e-3] [--max-iter-delta 10]
  compare_runs.py --no-state <log_a> <log_b> [options]

Exit code 0 = PASS, 1 = FAIL.
"""
import argparse
import re
import struct
import sys

FAMS = ("rmncc", "zmnsc", "lmnsc", "rmnss", "zmncs", "lmncs")

# Per-iteration row: "    0 | 1.616e-01   1.616e-01   4.775e-02 | 9.00e-01 ..."
ITER_RE = re.compile(
    r"^\s*(\d+)\s*\|\s*([0-9.eE+-]+)\s+([0-9.eE+-]+)\s+([0-9.eE+-]+)\s*\|\s*([0-9.eE+-]+)"
)
# Restart / progress markers
BADJ_RE = re.compile(r"BAD JACOBIAN \(iter2=(\d+)\) delt=([0-9.eE+-]+)")
BADP_RE = re.compile(r"BAD PROGRESS \(iter2=(\d+)\) delt=([0-9.eE+-]+)")
RESET_RE = re.compile(r"CONVERGENCE PROBLEM: RESETTING DELT to ([0-9.eE+-]+) \(ijacob=(\d+)\)")
CONV_RE = re.compile(r"CONVERGED at iter (\d+)")


def load_state(path):
    """Read the schema-v1 binary state payload: six mode-major families.

    Accepts on-disk container versions 1-8 (magic "CUMES001" + version +
    ns + mnmax header; the provenance trailer is ignored), plus the legacy
    pre-container raw dump the frozen baselines were captured in (int32 ns +
    int32 mnmax + the same six families, no magic, exactly sized).
    """
    fail = f"error: {path} is not a cumes state container"
    with open(path, "rb") as f:
        data = f.read()

    def families(offset, ns, mnmax):
        fams = {}
        pos = offset
        for name in FAMS:
            fams[name] = struct.unpack_from(f"<{ns * mnmax}d", data, pos)
            pos += 8 * ns * mnmax
        return fams

    header = struct.Struct("<8siii")  # magic(8) version ns mnmax = 20 bytes
    if len(data) >= header.size:
        magic, version, ns, mnmax = header.unpack_from(data)
        if magic == b"CUMES001" and 1 <= version <= 8 and ns >= 1 and mnmax >= 1:
            try:
                return ns, mnmax, families(header.size, ns, mnmax)
            except struct.error:
                raise SystemExit(fail)

    # Legacy raw baseline dump (pre-container): ns, mnmax, six families.
    if len(data) >= 8:
        ns, mnmax = struct.unpack_from("<ii", data)
        if ns >= 1 and mnmax >= 1 and len(data) == 8 + 6 * ns * mnmax * 8:
            return ns, mnmax, families(8, ns, mnmax)

    raise SystemExit(fail)


def parse_log(path):
    """Return (rows, restarts, converged_iter, status, iterations, summary).

    rows: dict iter -> (fsqr, fsqz, fsql, delt)
    restarts: list of (kind, iter2, delt); kind in {BADJ, BADP, RESET}
    summary: dict of "FSQR"/"FSQZ"/"FSQL" final values, or None
    """
    rows = {}
    restarts = []
    converged_iter = None
    status = None
    iterations = None
    summary = {}
    with open(path, "r") as f:
        for line in f:
            m = ITER_RE.match(line)
            if m:
                it = int(m.group(1))
                fr, fz, fl, dt = (float(m.group(i)) for i in (2, 3, 4, 5))
                rows[it] = (fr, fz, fl, dt)
                continue
            m = BADJ_RE.search(line)
            if m:
                restarts.append(("BADJ", int(m.group(1)), float(m.group(2))))
                continue
            m = BADP_RE.search(line)
            if m:
                restarts.append(("BADP", int(m.group(1)), float(m.group(2))))
                continue
            m = RESET_RE.search(line)
            if m:
                restarts.append(("RESET", None, float(m.group(1))))
                continue
            m = CONV_RE.search(line)
            if m:
                converged_iter = int(m.group(1))
                continue
            if line.startswith("  Status:"):
                status = line.split(":", 1)[1].strip()
            elif line.startswith("  Iterations:"):
                iterations = int(line.split(":", 1)[1].strip())
            elif line.startswith("  FSQR:"):
                summary["FSQR"] = float(line.split(":", 1)[1].strip())
            elif line.startswith("  FSQZ:"):
                summary["FSQZ"] = float(line.split(":", 1)[1].strip())
            elif line.startswith("  FSQL:"):
                summary["FSQL"] = float(line.split(":", 1)[1].strip())
    return rows, restarts, converged_iter, status, iterations, summary


def rel(a, b):
    """Difference scaled like compare_states.py: |a-b|/max(1, |b|).

    Effectively absolute for |b| < 1 (so tiny coefficients don't blow up the
    metric), relative for |b| > 1.
    """
    return abs(a - b) / max(1.0, abs(b))


def compare(rows_a, rows_b, tol):
    """Per-row residual comparison over the common iteration range."""
    iters = sorted(set(rows_a) & set(rows_b))
    only_a = sorted(set(rows_a) - set(rows_b))
    only_b = sorted(set(rows_b) - set(rows_a))
    worst = {k: 0.0 for k in ("fsqr", "fsqz", "fsql", "delt")}
    worst_at = {k: None for k in worst}
    for it in iters:
        a, b = rows_a[it], rows_b[it]
        for idx, k in enumerate(worst):
            d = rel(a[idx], b[idx])
            if d > worst[k]:
                worst[k], worst_at[k] = d, it
    ok = all(w < tol for w in worst.values()) and not only_a and not only_b
    return ok, worst, worst_at, len(iters), only_a, only_b


def compare_state(sa, sb, tol):
    (a_ns, a_mn, a), (b_ns, b_mn, b) = sa, sb
    if (a_ns, a_mn) != (b_ns, b_mn):
        return False, f"size mismatch: {(a_ns, a_mn)} vs {(b_ns, b_mn)}", {}
    ns, mnmax = a_ns, a_mn
    worst = {}
    for name in FAMS:
        amax = 0.0
        for m in range(mnmax):
            for j in range(1, ns):  # skip the extrapolated axis row
                i = m * ns + j
                amax = max(amax, rel(a[name][i], b[name][i]))
        worst[name] = amax
    ok = all(v < tol for v in worst.values())
    return ok, f"ns={ns} mnmax={mnmax} (interior j=1..{ns-1})", worst


def main():
    ap = argparse.ArgumentParser(description=__doc__,
                                 formatter_class=argparse.RawDescriptionHelpFormatter)
    ap.add_argument("log_a")
    ap.add_argument("state_a", nargs="?")
    ap.add_argument("log_b")
    ap.add_argument("state_b", nargs="?")
    ap.add_argument("--no-state", action="store_true", help="log-only comparison")
    ap.add_argument("--tol", type=float, default=1e-8, help="state tolerance (default 1e-8)")
    ap.add_argument("--log-tol", type=float, default=5e-3,
                    help="printed-row tolerance (default 5e-3; print precision is ~1e-4)")
    ap.add_argument("--max-iter-delta", type=int, default=10,
                    help="max |converged iters| difference (default 10)")
    args = ap.parse_args()

    if not args.no_state and (not args.state_a or not args.state_b):
        print("error: need state files, or pass --no-state for log-only comparison")
        return 2

    fails = []

    # ---- Per-iteration residuals + markers ----
    rows_a, rst_a, conv_a, stat_a, niter_a, summ_a = parse_log(args.log_a)
    rows_b, rst_b, conv_b, stat_b, niter_b, summ_b = parse_log(args.log_b)
    ok, worst, worst_at, n_rows, only_a, only_b = compare(rows_a, rows_b, args.log_tol)
    print(f"residual rows: {n_rows} iterations compared (worst rel diff per column)")
    for k in ("fsqr", "fsqz", "fsql", "delt"):
        print(f"  {k:4s}: {worst[k]:.3e} @ iter {worst_at[k]}" + ("  <-- EXCEEDS TOL" if worst[k] >= args.log_tol else ""))
    if only_a:
        print(f"  rows only in A: {only_a[:10]}{' ...' if len(only_a) > 10 else ''}")
    if only_b:
        print(f"  rows only in B: {only_b[:10]}{' ...' if len(only_b) > 10 else ''}")
    if not ok:
        fails.append("per-iteration residuals")

    # ---- Restart sequence ----
    seq_a = [(k, i) for k, i, _ in rst_a]
    seq_b = [(k, i) for k, i, _ in rst_b]
    print(f"restarts: A={len(rst_a)} events, B={len(rst_b)} events")
    if seq_a != seq_b:
        fails.append("restart sequence")
        print(f"  A: {seq_a}")
        print(f"  B: {seq_b}")
    else:
        print(f"  identical: {seq_a}")
        for (ka, ia, da), (kb, ib, db) in zip(rst_a, rst_b):
            if ka == "RESET":
                if rel(da, db) >= args.log_tol:
                    print(f"  RESET delt drift: {da:.3e} vs {db:.3e}")
                    fails.append("reset delt")
            elif rel(da, db) >= args.log_tol:
                print(f"  {ka}@iter{ia} delt drift: {da:.3e} vs {db:.3e}")
                fails.append("restart delt")

    # ---- Convergence ----
    print(f"converged: A={'iter ' + str(conv_a) if conv_a else 'NO'}  "
          f"B={'iter ' + str(conv_b) if conv_b else 'NO'}  (status {stat_a} / {stat_b})")
    if not conv_a or not conv_b:
        fails.append("one or both runs did not converge")
    else:
        d = abs(conv_a - conv_b)
        print(f"  converged-iter delta: {d} (window {args.max_iter_delta})")
        if d > args.max_iter_delta:
            fails.append("converged iteration count")
    for k in ("FSQR", "FSQZ", "FSQL"):
        if k in summ_a and k in summ_b:
            d = rel(summ_a[k], summ_b[k])
            print(f"  final {k}: A={summ_a[k]:.3e} B={summ_b[k]:.3e} rel={d:.2e}")
            if d >= args.log_tol:
                fails.append(f"final {k}")

    # ---- State ----
    if not args.no_state:
        sa, sb = load_state(args.state_a), load_state(args.state_b)
        ok_s, msg, worst_s = compare_state(sa, sb, args.tol)
        print(f"state: {msg}")
        for name in FAMS:
            print(f"  {name:6s}: max rel diff (interior) = {worst_s[name]:.3e}"
                  + ("  <-- EXCEEDS TOL" if worst_s[name] >= args.tol else ""))
        if not ok_s:
            fails.append("state")

    if fails:
        print(f"FAIL: {', '.join(fails)}")
        return 1
    print("PASS")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/compare_runs.cpp
// Compare two cuMES trajectories and, optionally, their converged states.

#include "../include/clap.h"
#include "include/compare_common.hpp"

#include <algorithm>
#include <array>
#include <cerrno>
#include <cmath>
#include <cstdlib>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <map>
#include <optional>
#include <regex>
#include <set>
#include <sstream>
#include <stdexcept>
#include <string>
#include <tuple>
#include <vector>

namespace {

struct CommandLine {
    std::string log_a;
    std::string state_a;
    std::string log_b;
    std::string state_b;
    bool no_state = false;
    double tolerance = 1.0e-8;
    double log_tolerance = 5.0e-3;
    int max_iteration_delta = 10;
};

struct WithStateArguments {
    std::string log_a;
    std::string state_a;
    std::string log_b;
    std::string state_b;
    bool no_state = false;
    double tolerance = 1.0e-8;
    double log_tolerance = 5.0e-3;
    int max_iteration_delta = 10;
};

struct NoStateArguments {
    std::string log_a;
    std::string log_b;
    bool no_state = false;
    double tolerance = 1.0e-8;
    double log_tolerance = 5.0e-3;
    int max_iteration_delta = 10;
};

int parse_with_state(WithStateArguments& input, int argc, char** argv) {
    CLAP_BEGIN(WithStateArguments)
    CLAP_ADD_USAGE(
        "LOG_A STATE_A LOG_B STATE_B [--tol X] [--log-tol X] "
        "[--max-iter-delta N]")
    CLAP_ADD_DESCRIPTION(
        "Compare residual logs, restart decisions, convergence, and state.")
    CLAP_REGISTER_ARG(log_a)
    CLAP_REGISTER_ARG(state_a)
    CLAP_REGISTER_ARG(log_b)
    CLAP_REGISTER_ARG(state_b)
    CLAP_REGISTER_OPTION_WITH_DESCRIPTION(
        no_state, "--no-state", "compare two logs without state files")
    CLAP_REGISTER_OPTION_WITH_DESCRIPTION(
        tolerance, "--tol", "state relative tolerance (default 1e-8)")
    CLAP_REGISTER_OPTION_WITH_DESCRIPTION(
        log_tolerance, "--log-tol", "printed residual tolerance (default 5e-3)")
    CLAP_REGISTER_OPTION_WITH_DESCRIPTION(
        max_iteration_delta, "--max-iter-delta",
        "maximum convergence-iteration difference (default 10)")
    CLAP_END(WithStateArguments)
    try {
        CLAP<WithStateArguments>::parse_input(input, argc, argv);
    } catch (const std::exception& error) {
        std::cerr << error.what();
        return EINVAL;
    }
    return 0;
}

int parse_without_state(NoStateArguments& input, int argc, char** argv) {
    CLAP_BEGIN(NoStateArguments)
    CLAP_ADD_USAGE("--no-state LOG_A LOG_B [--log-tol X] [--max-iter-delta N]")
    CLAP_ADD_DESCRIPTION(
        "Compare residual logs, restart decisions, and convergence.")
    CLAP_REGISTER_ARG(log_a)
    CLAP_REGISTER_ARG(log_b)
    CLAP_REGISTER_OPTION_WITH_DESCRIPTION(
        no_state, "--no-state", "compare two logs without state files")
    CLAP_REGISTER_OPTION_WITH_DESCRIPTION(
        tolerance, "--tol", "unused state tolerance (accepted for parity)")
    CLAP_REGISTER_OPTION_WITH_DESCRIPTION(
        log_tolerance, "--log-tol", "printed residual tolerance (default 5e-3)")
    CLAP_REGISTER_OPTION_WITH_DESCRIPTION(
        max_iteration_delta, "--max-iter-delta",
        "maximum convergence-iteration difference (default 10)")
    CLAP_END(NoStateArguments)
    try {
        CLAP<NoStateArguments>::parse_input(input, argc, argv);
    } catch (const std::exception& error) {
        std::cerr << error.what();
        return EINVAL;
    }
    return 0;
}

int parse_command_line(CommandLine& command, int argc, char** argv) {
    bool requested_no_state = false;
    for (int index = 1; index < argc; ++index) {
        if (std::string(argv[index]) == "--no-state") {
            requested_no_state = true;
            break;
        }
    }
    if (requested_no_state) {
        NoStateArguments input;
        if (const int status = parse_without_state(input, argc, argv); status) {
            return status;
        }
        command.log_a = std::move(input.log_a);
        command.log_b = std::move(input.log_b);
        command.no_state = true;
        command.tolerance = input.tolerance;
        command.log_tolerance = input.log_tolerance;
        command.max_iteration_delta = input.max_iteration_delta;
    } else {
        WithStateArguments input;
        if (const int status = parse_with_state(input, argc, argv); status) {
            return status;
        }
        command.log_a = std::move(input.log_a);
        command.state_a = std::move(input.state_a);
        command.log_b = std::move(input.log_b);
        command.state_b = std::move(input.state_b);
        command.no_state = false;
        command.tolerance = input.tolerance;
        command.log_tolerance = input.log_tolerance;
        command.max_iteration_delta = input.max_iteration_delta;
    }
    if (!(command.tolerance > 0.0) || !std::isfinite(command.tolerance) ||
        !(command.log_tolerance > 0.0) ||
        !std::isfinite(command.log_tolerance) ||
        command.max_iteration_delta < 0) {
        std::cerr << "error: tolerances must be finite and positive and "
                     "--max-iter-delta must be nonnegative\n";
        return EINVAL;
    }
    return 0;
}

using Residual = std::array<double, 4>;

struct Restart {
    std::string kind;
    std::optional<int> iteration;
    double delt = 0.0;
};

struct LogData {
    std::map<int, Residual> rows;
    std::vector<Restart> restarts;
    std::optional<int> converged_iteration;
    std::string status;
    std::optional<int> iterations;
    std::map<std::string, double> summary;
};

std::string trim(std::string text) {
    const auto first = text.find_first_not_of(" \t\r\n");
    if (first == std::string::npos) return {};
    const auto last = text.find_last_not_of(" \t\r\n");
    return text.substr(first, last - first + 1);
}

bool starts_with(const std::string& text, const std::string& prefix) {
    return text.size() >= prefix.size() &&
           text.compare(0, prefix.size(), prefix) == 0;
}

std::string after_colon(const std::string& line) {
    const auto colon = line.find(':');
    return colon == std::string::npos ? std::string{}
                                      : trim(line.substr(colon + 1));
}

LogData parse_log(const std::string& path) {
    std::ifstream stream(path);
    if (!stream) throw std::runtime_error("could not open " + path);

    const std::regex iteration(
        R"(^\s*(\d+)\s*\|\s*([0-9.eE+\-]+)\s+([0-9.eE+\-]+)\s+([0-9.eE+\-]+)\s*\|\s*([0-9.eE+\-]+))");
    const std::regex bad_jacobian(
        R"(BAD JACOBIAN \(iter2=(\d+)\) delt=([0-9.eE+\-]+))");
    const std::regex bad_progress(
        R"(BAD PROGRESS \(iter2=(\d+)\) delt=([0-9.eE+\-]+))");
    const std::regex reset(
        R"(CONVERGENCE PROBLEM: RESETTING DELT to ([0-9.eE+\-]+) \(ijacob=(\d+)\))");
    const std::regex converged(R"(CONVERGED at iter (\d+))");

    LogData result;
    std::string line;
    std::smatch match;
    while (std::getline(stream, line)) {
        if (std::regex_search(line, match, iteration)) {
            const int index = std::stoi(match[1].str());
            result.rows[index] = {
                std::stod(match[2].str()), std::stod(match[3].str()),
                std::stod(match[4].str()), std::stod(match[5].str())};
        } else if (std::regex_search(line, match, bad_jacobian)) {
            result.restarts.push_back(
                {"BADJ", std::stoi(match[1].str()), std::stod(match[2].str())});
        } else if (std::regex_search(line, match, bad_progress)) {
            result.restarts.push_back(
                {"BADP", std::stoi(match[1].str()), std::stod(match[2].str())});
        } else if (std::regex_search(line, match, reset)) {
            result.restarts.push_back(
                {"RESET", std::nullopt, std::stod(match[1].str())});
        } else if (std::regex_search(line, match, converged)) {
            result.converged_iteration = std::stoi(match[1].str());
        } else if (starts_with(line, "  Status:")) {
            result.status = after_colon(line);
        } else if (starts_with(line, "  Iterations:")) {
            result.iterations = std::stoi(after_colon(line));
        } else if (starts_with(line, "  FSQR:")) {
            result.summary["FSQR"] = std::stod(after_colon(line));
        } else if (starts_with(line, "  FSQZ:")) {
            result.summary["FSQZ"] = std::stod(after_colon(line));
        } else if (starts_with(line, "  FSQL:")) {
            result.summary["FSQL"] = std::stod(after_colon(line));
        }
    }
    return result;
}

struct RowComparison {
    bool pass = false;
    std::array<double, 4> worst{};
    std::array<std::optional<int>, 4> worst_at{};
    std::size_t common_count = 0;
    std::vector<int> only_a;
    std::vector<int> only_b;
};

RowComparison compare_rows(const std::map<int, Residual>& a,
                           const std::map<int, Residual>& b,
                           double tolerance) {
    RowComparison result;
    for (const auto& [iteration, row] : a) {
        const auto found = b.find(iteration);
        if (found == b.end()) {
            result.only_a.push_back(iteration);
            continue;
        }
        ++result.common_count;
        for (std::size_t column = 0; column < result.worst.size(); ++column) {
            const double difference = cumes::compare::relative_difference(
                row[column], found->second[column]);
            if (difference > result.worst[column]) {
                result.worst[column] = difference;
                result.worst_at[column] = iteration;
            }
        }
    }
    for (const auto& [iteration, unused] : b) {
        (void)unused;
        if (a.find(iteration) == a.end()) {
            result.only_b.push_back(iteration);
        }
    }
    result.pass =
        result.only_a.empty() && result.only_b.empty() &&
        std::all_of(result.worst.begin(), result.worst.end(),
                    [tolerance](double value) { return value < tolerance; });
    return result;
}

void print_iteration_list(const std::vector<int>& values) {
    std::cout << '[';
    const auto shown = std::min<std::size_t>(10, values.size());
    for (std::size_t index = 0; index < shown; ++index) {
        if (index) std::cout << ", ";
        std::cout << values[index];
    }
    if (values.size() > shown) std::cout << " ...";
    std::cout << ']';
}

bool same_restart_sequence(const std::vector<Restart>& a,
                           const std::vector<Restart>& b) {
    if (a.size() != b.size()) return false;
    for (std::size_t index = 0; index < a.size(); ++index) {
        if (a[index].kind != b[index].kind ||
            a[index].iteration != b[index].iteration) {
            return false;
        }
    }
    return true;
}

void print_restart_sequence(const std::vector<Restart>& events) {
    std::cout << '[';
    for (std::size_t index = 0; index < events.size(); ++index) {
        if (index) std::cout << ", ";
        std::cout << '(' << events[index].kind << ", ";
        if (events[index].iteration) {
            std::cout << *events[index].iteration;
        } else {
            std::cout << "None";
        }
        std::cout << ')';
    }
    std::cout << ']';
}

struct StateComparison {
    bool pass = false;
    std::string dimensions;
    std::array<double, cumes::compare::FAMILY_NAMES.size()> worst{};
};

StateComparison compare_states(const cumes::compare::State& a,
                               const cumes::compare::State& b,
                               double tolerance) {
    StateComparison result;
    if (a.ns != b.ns || a.mnmax != b.mnmax) {
        std::ostringstream message;
        message << "size mismatch: (" << a.ns << ", " << a.mnmax << ") vs ("
                << b.ns << ", " << b.mnmax << ')';
        result.dimensions = message.str();
        return result;
    }
    std::ostringstream message;
    message << "ns=" << a.ns << " mnmax=" << a.mnmax << " (interior j=1.."
            << a.ns - 1 << ')';
    result.dimensions = message.str();
    result.pass = true;
    for (std::size_t family = 0; family < result.worst.size(); ++family) {
        for (std::int32_t mode = 0; mode < a.mnmax; ++mode) {
            for (std::int32_t surface = 1; surface < a.ns; ++surface) {
                const auto index = static_cast<std::size_t>(mode) *
                                       static_cast<std::size_t>(a.ns) +
                                   static_cast<std::size_t>(surface);
                result.worst[family] = std::max(
                    result.worst[family],
                    cumes::compare::relative_difference(
                        a.families[family][index], b.families[family][index]));
            }
        }
        result.pass = result.pass && result.worst[family] < tolerance;
    }
    return result;
}

void append_failure(std::vector<std::string>& failures,
                    const std::string& failure) {
    failures.push_back(failure);
}

std::string optional_iteration(const std::optional<int>& value) {
    return value ? std::to_string(*value) : "None";
}

}  // namespace

int main(int argc, char** argv) {
    CommandLine command;
    if (const int status = parse_command_line(command, argc, argv); status) {
        return status;
    }

    try {
        const auto a = parse_log(command.log_a);
        const auto b = parse_log(command.log_b);
        std::vector<std::string> failures;

        const auto rows = compare_rows(a.rows, b.rows, command.log_tolerance);
        constexpr std::array<const char*, 4> columns = {"fsqr", "fsqz", "fsql",
                                                        "delt"};
        std::cout << "residual rows: " << rows.common_count
                  << " iterations compared (worst rel diff per column)\n";
        for (std::size_t column = 0; column < columns.size(); ++column) {
            std::cout << "  " << std::left << std::setw(4) << columns[column]
                      << std::right << ": " << std::scientific
                      << std::setprecision(3) << rows.worst[column]
                      << " @ iter "
                      << optional_iteration(rows.worst_at[column]);
            if (rows.worst[column] >= command.log_tolerance) {
                std::cout << "  <-- EXCEEDS TOL";
            }
            std::cout << '\n';
        }
        if (!rows.only_a.empty()) {
            std::cout << "  rows only in A: ";
            print_iteration_list(rows.only_a);
            std::cout << '\n';
        }
        if (!rows.only_b.empty()) {
            std::cout << "  rows only in B: ";
            print_iteration_list(rows.only_b);
            std::cout << '\n';
        }
        if (!rows.pass) append_failure(failures, "per-iteration residuals");

        std::cout << "restarts: A=" << a.restarts.size()
                  << " events, B=" << b.restarts.size() << " events\n";
        if (!same_restart_sequence(a.restarts, b.restarts)) {
            append_failure(failures, "restart sequence");
            std::cout << "  A: ";
            print_restart_sequence(a.restarts);
            std::cout << "\n  B: ";
            print_restart_sequence(b.restarts);
            std::cout << '\n';
        } else {
            std::cout << "  identical: ";
            print_restart_sequence(a.restarts);
            std::cout << '\n';
            for (std::size_t index = 0; index < a.restarts.size(); ++index) {
                const auto& left = a.restarts[index];
                const auto& right = b.restarts[index];
                if (cumes::compare::relative_difference(
                        left.delt, right.delt) >= command.log_tolerance) {
                    if (left.kind == "RESET") {
                        std::cout << "  RESET delt drift: " << std::scientific
                                  << std::setprecision(3) << left.delt << " vs "
                                  << right.delt << '\n';
                        append_failure(failures, "reset delt");
                    } else {
                        std::cout << "  " << left.kind << "@iter"
                                  << *left.iteration
                                  << " delt drift: " << std::scientific
                                  << std::setprecision(3) << left.delt << " vs "
                                  << right.delt << '\n';
                        append_failure(failures, "restart delt");
                    }
                }
            }
        }

        std::cout << "converged: A=";
        if (a.converged_iteration) {
            std::cout << "iter " << *a.converged_iteration;
        } else {
            std::cout << "NO";
        }
        std::cout << "  B=";
        if (b.converged_iteration) {
            std::cout << "iter " << *b.converged_iteration;
        } else {
            std::cout << "NO";
        }
        std::cout << "  (status " << (a.status.empty() ? "None" : a.status)
                  << " / " << (b.status.empty() ? "None" : b.status) << ")\n";
        if (!a.converged_iteration || !b.converged_iteration) {
            append_failure(failures, "one or both runs did not converge");
        } else {
            const int difference =
                std::abs(*a.converged_iteration - *b.converged_iteration);
            std::cout << "  converged-iter delta: " << difference << " (window "
                      << command.max_iteration_delta << ")\n";
            if (difference > command.max_iteration_delta) {
                append_failure(failures, "converged iteration count");
            }
        }
        for (const char* key : {"FSQR", "FSQZ", "FSQL"}) {
            const auto left = a.summary.find(key);
            const auto right = b.summary.find(key);
            if (left == a.summary.end() || right == b.summary.end()) continue;
            const double difference = cumes::compare::relative_difference(
                left->second, right->second);
            std::cout << "  final " << key << ": A=" << std::scientific
                      << std::setprecision(3) << left->second
                      << " B=" << right->second
                      << " rel=" << std::setprecision(2) << difference << '\n';
            if (difference >= command.log_tolerance) {
                append_failure(failures, std::string("final ") + key);
            }
        }

        if (!command.no_state) {
            const auto left = cumes::compare::read_state(command.state_a, true);
            const auto right =
                cumes::compare::read_state(command.state_b, true);
            const auto state = compare_states(left, right, command.tolerance);
            std::cout << "state: " << state.dimensions << '\n';
            if (left.ns == right.ns && left.mnmax == right.mnmax) {
                for (std::size_t family = 0; family < state.worst.size();
                     ++family) {
                    std::cout
                        << "  " << std::left << std::setw(6)
                        << cumes::compare::FAMILY_NAMES[family] << std::right
                        << ": max rel diff (interior) = " << std::scientific
                        << std::setprecision(3) << state.worst[family];
                    if (state.worst[family] >= command.tolerance) {
                        std::cout << "  <-- EXCEEDS TOL";
                    }
                    std::cout << '\n';
                }
            }
            if (!state.pass) append_failure(failures, "state");
        }

        if (!failures.empty()) {
            std::cout << "FAIL: ";
            for (std::size_t index = 0; index < failures.size(); ++index) {
                if (index) std::cout << ", ";
                std::cout << failures[index];
            }
            std::cout << '\n';
            return EXIT_FAILURE;
        }
        std::cout << "PASS\n";
        return EXIT_SUCCESS;
    } catch (const std::exception& error) {
        std::cerr << "error: " << error.what() << '\n';
        return 2;
    }
}
